catalog/permissions/permissions.go

```diff
@@ -0,0 +1,152 @@
+	return retry.RetryContext(a.context, timeout, func() *retry.RetryError {
+		current, err := a.GetPermissions(securable, name)
+		if err != nil {
+			return retry.NonRetryableError(err)
+		}
+		log.Printf("[DEBUG] Permissions for %s-%s are: %v", securable.String(), name, current)
+		if diff(current, desired) == nil {
+			return nil
+		}
+		return retry.RetryableError(
+			fmt.Errorf("permissions for %s-%s are %v, but have to be %v", securable.String(), name, current, desired),
+		)
+	})
+}
+
+// Terraform Schema
+type UnityCatalogPrivilegeAssignment struct {
+	Principal  string   `json:"principal"`
+	Privileges []string `json:"privileges" tf:"slice_set"`
+}
+
+// Permission Mappings
+
+type SecurableMapping map[string]catalog.SecurableType
+
+// reuse ResourceDiff and ResourceData
+type attributeGetter interface {
+	Get(key string) any
+}
+
+func (sm SecurableMapping) GetSecurableType(securable string) catalog.SecurableType {
+	return sm[securable]
+}
+
+func (sm SecurableMapping) KeyValue(d attributeGetter) (string, string) {
+	for field := range sm {
+		v := d.Get(field).(string)
+		if v == "" {
+			continue
+		}
+		return field, v
+	}
+	return "unknown", "unknown"
+}
+func (sm SecurableMapping) Id(d *schema.ResourceData) string {
+	securable, name := sm.KeyValue(d)
+	return fmt.Sprintf("%s/%s", securable, name)
+}
+
+// Mappings
+// See https://docs.databricks.com/api/workspace/grants/update for full list
+// Omitting provider as a reserved keyword
+var Mappings = SecurableMapping{
+	"catalog":            catalog.SecurableType("catalog"),
+	"foreign_connection": catalog.SecurableType("connection"),
+	"external_location":  catalog.SecurableType("external_location"),
+	"function":           catalog.SecurableType("function"),
+	"metastore":          catalog.SecurableType("metastore"),
+	"model":              catalog.SecurableType("function"),
+	"pipeline":           catalog.SecurableType("pipeline"),
+	"recipient":          catalog.SecurableType("recipient"),
+	"schema":             catalog.SecurableType("schema"),
+	"share":              catalog.SecurableType("share"),
+	"storage_credential": catalog.SecurableType("storage_credential"),
+	"table":              catalog.SecurableType("table"),
+	"volume":             catalog.SecurableType("volume"),
+}
+
+// Utils for Slice and Set
+func SliceToSet(in []catalog.Privilege) *schema.Set {
+	var out []any
+	for _, v := range in {
+		out = append(out, v.String())
+	}
+	return schema.NewSet(schema.HashString, out)
+}
+
+func SetToSlice(set *schema.Set) (ss []catalog.Privilege) {
+	for _, v := range set.List() {
+		ss = append(ss, catalog.Privilege(v.(string)))
+	}
+	return
+}
+
+func ToPrivilegeSlice(in []interface{}) (out []catalog.Privilege) {
+	for _, v := range in {
+		out = append(out, catalog.Privilege(v.(string)))
+	}
+	return
+}
+
+func SliceWithoutString(in []string, without string) (out []string) {
```

Comment:
This is the same as `slices.DeleteFunc(in, func(s string) { return s == without })`. Can we use the standard library function instead?

Reply:
We could but, DeleteFunc modifies the original slice & the tests fail without cloning the slice first due to side effects. I can swap the code to the following if you prefer:
```suggestion
func SliceWithoutString(in []string, without string) (out []string) {
	out = slices.Clone(in)
	out = slices.DeleteFunc(out, func(s string) bool { return s == without })
	return
}

```

Reply:
Ah, good point. You can leave this as-is then.